File: core/engines/deepgram.py
"""Deepgram Nova-3 어댑터.

요금: nova-3 Pay-as-you-go $0.0043/분 기준 (2026-04 공개가).
파일은 바이트 직접 업로드 (2GB 한도).
"""

from __future__ import annotations

import os
import time
from pathlib import Path
from typing import Any

import httpx

from .base import TranscriptResult, Utterance, Word

DEEPGRAM_URL = "https://api.deepgram.com/v1/listen"
COST_PER_MINUTE_USD = 0.0043
MAX_RETRIES = 3
BASE_BACKOFF_SEC = 2.0
REQUEST_TIMEOUT_SEC = 1800.0  # 30분 — 큰 파일 업로드 여유


class DeepgramError(RuntimeError):
    pass


class DeepgramEngine:
    name = "deepgram"
    model = "nova-3"

    def __init__(self, api_key: str | None = None, model: str | None = None) -> None:
        key = api_key or os.environ.get("DEEPGRAM_API_KEY")
        if not key:
            raise DeepgramError("DEEPGRAM_API_KEY 가 설정되지 않았습니다.")
        self._key = key
        if model:
            self.model = model

    def estimate_cost_usd(self, duration_sec: float) -> float:
        return (duration_sec / 60.0) * COST_PER_MINUTE_USD

    def transcribe(self, audio_path: Path, language: str) -> TranscriptResult:
        payload = audio_path.read_bytes()
        params = {
            "model": self.model,
            "smart_format": "true",
            "punctuate": "true",
            "utterances": "true",
        }
        if language and language != "auto":
            params["language"] = language
        else:
            params["detect_language"] = "true"

        content_type = _mime_for(audio_path)
        data = _post_with_retry(self._key, params, payload, content_type)
        return _parse(data, model=self.model)


def _post_with_retry(
    api_key: str,
    params: dict[str, str],
    payload: bytes,
    content_type: str,
) -> dict[str, Any]:
    last_err: Exception | None = None
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            with httpx.Client(timeout=REQUEST_TIMEOUT_SEC) as client:
                resp = client.post(
                    DEEPGRAM_URL,
                    params=params,
                    content=payload,
                    headers={
                        "Authorization": f"Token {api_key}",
                        "Content-Type": content_type,
                    },
                )
            if resp.status_code == 200:
                return resp.json()
            if resp.status_code in (429, 500, 502, 503, 504):
                raise DeepgramError(f"Deepgram {resp.status_code}: {resp.text[:300]}")
            raise DeepgramError(f"Deepgram {resp.status_code}: {resp.text[:300]}")
        except (httpx.HTTPError, DeepgramError) as e:
            last_err = e
            if attempt == MAX_RETRIES:
                break
            time.sleep(BASE_BACKOFF_SEC * (2 ** (attempt - 1)))
    raise DeepgramError(f"Deepgram 요청 최종 실패: {last_err}")


def _mime_for(path: Path) -> str:
    mapping = {
        ".m4a": "audio/mp4",
        ".mp4": "audio/mp4",
        ".aac": "audio/aac",
        ".mp3": "audio/mpeg",
        ".wav": "audio/wav",
        ".flac": "audio/flac",
        ".ogg": "audio/ogg",
        ".opus": "audio/ogg",
        ".webm": "audio/webm",
    }
    return mapping.get(path.suffix.lower(), "application/octet-stream")


def _parse(data: dict[str, Any], *, model: str) -> TranscriptResult:
    metadata = data.get("metadata", {}) or {}
    results = data.get("results", {}) or {}
    channels = results.get("channels", []) or []
    alt = (channels[0].get("alternatives", [{}])[0] if channels else {}) or {}

    text = alt.get("transcript", "") or ""
    raw_words = alt.get("words", []) or []
    words = tuple(
        Word(
            text=w.get("punctuated_word") or w.get("word", ""),
            start=float(w.get("start", 0.0)),
            end=float(w.get("end", 0.0)),
            confidence=_maybe_float(w.get("confidence")),
            speaker=_maybe_int(w.get("speaker")),
        )
        for w in raw_words
    )

    raw_utts = results.get("utterances", []) or []
    utterances = tuple(
        Utterance(
            text=u.get("transcript", "") or "",
            start=float(u.get("start", 0.0)),
            end=float(u.get("end", 0.0)),
            speaker=_maybe_int(u.get("speaker")),
            words=tuple(
                Word(
                    text=w.get("punctuated_word") or w.get("word", ""),
                    start=float(w.get("start", 0.0)),
                    end=float(w.get("end", 0.0)),
                    confidence=_maybe_float(w.get("confidence")),
                    speaker=_maybe_int(w.get("speaker")),
                )
                for w in (u.get("words", []) or [])
            ),
        )
        for u in raw_utts
    )

    language = _detect_language(alt, metadata)
    duration = float(metadata.get("duration", 0.0) or 0.0)

    return TranscriptResult(
        text=text,
        words=words,
        utterances=utterances,
        language=language,
        engine="deepgram",
        model=model,
        duration_sec=duration,
        raw=data,
    )


def _detect_language(alt: dict, metadata: dict) -> str:
    for key in ("detected_language", "language"):
        val = alt.get(key) or metadata.get(key)
        if val:
            return str(val)
    return ""


def _maybe_float(v: Any) -> float | None:
    if v is None:
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


def _maybe_int(v: Any) -> int | None:
    if v is None:
        return None
    try:
        return int(v)
    except (TypeError, ValueError):
        return None

File: web/lib/deepgram.ts
export const DEEPGRAM_COST_PER_MIN = 0.0043;
export const DEEPGRAM_MODEL = "nova-3";

export interface DeepgramWord {
  word?: string;
  punctuated_word?: string;
  start?: number;
  end?: number;
  confidence?: number;
}

export interface DeepgramUtterance {
  start?: number;
  end?: number;
  transcript?: string;
  words?: DeepgramWord[];
}

export type TranscribeLanguage = "auto" | "ko" | "en";

export function buildDeepgramAsyncUrl(
  callbackUrl: string,
  opts?: { language?: TranscribeLanguage },
): string {
  const params = new URLSearchParams({
    model: DEEPGRAM_MODEL,
    smart_format: "true",
    punctuate: "true",
    utterances: "true",
    diarize: "false",
    callback: callbackUrl,
  });
  const lang = opts?.language ?? "auto";
  if (lang === "auto") {
    params.set("detect_language", "true");
  } else {
    params.set("language", lang);
  }
  return `https://api.deepgram.com/v1/listen?${params.toString()}`;
}

export function extractTranscript(payload: unknown): {
  text: string;
  utterances: DeepgramUtterance[];
  words: DeepgramWord[];
  duration: number;
} {
  const p = payload as {
    results?: {
      channels?: { alternatives?: { transcript?: string; words?: DeepgramWord[] }[] }[];
      utterances?: DeepgramUtterance[];
    };
    metadata?: { duration?: number };
  };
  const alt = p?.results?.channels?.[0]?.alternatives?.[0];
  return {
    text: alt?.transcript ?? "",
    utterances: p?.results?.utterances ?? [],
    words: alt?.words ?? [],
    duration: p?.metadata?.duration ?? 0,
  };
}

export function buildTxt(utterances: DeepgramUtterance[], fallback: string): string {
  if (utterances.length) {
    return utterances.map((u) => (u.transcript ?? "").trim()).filter(Boolean).join("\n\n") + "\n";
  }
  return fallback.trim() + "\n";
}

export function buildSrt(utterances: DeepgramUtterance[]): string {
  const blocks: string[] = [];
  utterances.forEach((u, i) => {
    const text = (u.transcript ?? "").trim();
    if (!text) return;
    blocks.push(
      `${i + 1}\n${srtTs(u.start ?? 0)} --> ${srtTs(u.end ?? 0)}\n${text}\n`,
    );
  });
  return blocks.join("\n");
}

export function buildVtt(utterances: DeepgramUtterance[]): string {
  const lines = ["WEBVTT", ""];
  utterances.forEach((u) => {
    const text = (u.transcript ?? "").trim();
    if (!text) return;
    lines.push(`${vttTs(u.start ?? 0)} --> ${vttTs(u.end ?? 0)}`);
    lines.push(text);
    lines.push("");
  });
  return lines.join("\n");
}

function srtTs(sec: number): string {
  const ms = Math.round(sec * 1000);
  const h = Math.floor(ms / 3_600_000);
  const m = Math.floor((ms % 3_600_000) / 60_000);
  const s = Math.floor((ms % 60_000) / 1000);
  const r = ms % 1000;
  return `${pad(h, 2)}:${pad(m, 2)}:${pad(s, 2)},${pad(r, 3)}`;
}
function vttTs(sec: number): string {
  return srtTs(sec).replace(",", ".");
}
function pad(n: number, w: number): string {
  return String(n).padStart(w, "0");
}
